### PR TITLE
CDAP-3224 add api for creating an app from an artifact

### PR DESCRIPTION
**Pull request by albertshau**
Modifies the current PUT /apps/<appid> api so that requests with
a json content type are interpreted as a request to create an app
based on an artifact that already exists.

### CODE REVIEW

**Comment by awholegunch** [comment]
Should there be some documentation changes to reflect this, and indicate how backwards-compatibilty is handled?

**Comment by albertshau** [comment]
there will be doc changes, but wanted to do them separately

**Comment by albertshau** [comment]
http://builds.cask.co/browse/CDAP-DUT2479

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java [hunk 1]
```diff
@@ -244,6 +265,54 @@ public void deleteAllApps(HttpRequest request, HttpResponder responder,
     responder.sendString(status.getCode(), status.getMessage());
   }
 
+  private BodyConsumer deployAppFromArtifact(final Id.Namespace namespace, final String appId) throws IOException {
+
+    return new AbstractBodyConsumer(File.createTempFile(appId, ".json", tmpDir)) {
```
Comment (gokulavasan) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: Do we need BodyConsumer for json file that contains artifact info and configuration?

Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/AppLifecycleHttpHandler.java]: normally no, but we decided to do this weird thing where the path is the same as the current api which does need a body consumer, so the two need to have the same method signature.

File: cdap-proto/src/main/java/co/cask/cdap/proto/artifact/CreateAppRequest.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+/**
+ * Request body when creating an app
+ *
+ * @param <T> the type of application config
+ */
+public class CreateAppRequest<T> {
+  private final ArtifactSummary artifact;
+  private final T config;
+
+  public CreateAppRequest(ArtifactSummary artifact, T config) {
```
Comment (gokulavasan) [inline cdap-proto/src/main/java/co/cask/cdap/proto/artifact/CreateAppRequest.java]: Would it be useful to have a constructor that takes in only artifact summary?

**Comment by gokulavasan** [comment]
Couple of comments

**Comment by gokulavasan** [comment]
LGTM :+1:

**Pull request merged**